File: src/binds/c/instance/instance.c
#include <GLFW/glfw3.h>
#include <vulkan/vulkan.h>
#include <stdlib.h>

typedef struct str_arr {
	int len;
	const char** arr;
} StrArr;

static StrArr _wrangle_extensions(StrArr required_extensions) {
	uint32_t extension_count = 0;
	const char** extensions = glfwGetRequiredInstanceExtensions(&extension_count);
	StrArr combined;
	combined.arr = malloc((extension_count + required_extensions.len) * sizeof(const char*));
	for (int i = 0; i < extension_count + required_extensions.len; i++) {
		if (i < extension_count) {
			combined.arr[i] = extensions[i];
		}
		else {
			combined.arr[i] = required_extensions.arr[i-extension_count];
		}
	}
	combined.len = extension_count + required_extensions.len;
	return combined;
}

typedef struct instanceInitResult {
	VkResult result;
	VkInstance instance;
} instanceInitResult;

instanceInitResult instanceInit(
		const char* app_name, int major, int minor, int patch, 
		const char** required_extensions, int num_extensions, int flags,
		const char** validation_layers, int num_layers
	) {
	StrArr reqd_exts = { .len = num_extensions, .arr = required_extensions };
	StrArr vald_lays = { .len = num_layers, .arr = validation_layers };
	VkApplicationInfo app_info = {
		.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
		.pApplicationName = app_name,
		.applicationVersion = VK_MAKE_VERSION(major, minor, patch),
		.pEngineName = "No Engine",
		.engineVersion = VK_MAKE_VERSION(major, minor, patch),
		.apiVersion = VK_API_VERSION_1_0
	};

	StrArr extensions = _wrangle_extensions(reqd_exts);
	VkInstanceCreateInfo create_info = {
		.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
		.pApplicationInfo = &app_info,
		.enabledExtensionCount = extensions.len,
		.ppEnabledExtensionNames = extensions.arr,
		.enabledLayerCount = vald_lays.len,
		.ppEnabledLayerNames = vald_lays.arr,
		.flags = flags
	};

	instanceInitResult result;
	result.result = vkCreateInstance(&create_info, NULL, &result.instance);
	free(extensions.arr);
	return result;
}

void instanceDeinit(const instanceInitResult* instance) {
	vkDestroyInstance(instance->instance, NULL);
}

File: src/binds/c/instance/instance.zig
const glfw = @import("../glfw.zig");
const vk = @import("../vulkan.zig");
const std = @import("std");
const types = @import("../types.zig");

const c_str = [*c]const u8;
const str = []const u8;
const instanceInitResult = extern struct {
    result: vk.VkResult, 
    instance: vk.VkInstance
};
extern fn instanceInit(c_str, c_int, c_int, c_int, [*c]const c_str, c_int, c_int, [*c]const c_str, c_int) instanceInitResult;
extern fn instanceDeinit(*const instanceInitResult) void;

pub const errors = error{
    instanceCreationFailed
};

pub const instance = struct {
    inst: vk.VkInstance,
    result: instanceInitResult,

    pub fn init(alloc: std.mem.Allocator, app_name: str, version: [3]u32, extensions: []str, flags: i32, validations: []str) !instance {
        var inst: instance = undefined;
        try inst._init(alloc, app_name, version, extensions, flags, validations);
        return inst;
    }

    fn _init(self: *instance, alloc: std.mem.Allocator, 
            app_name: str, version: [3]u32, 
            required_extensions: []str, flags: i32,
            validation_layers: []str) !void {

        const z_app_name = try types.CStr.init(alloc, app_name);
        defer z_app_name.deinit();
        const z_extensions = try types.CStrArray.init(alloc, required_extensions);
        defer z_extensions.deinit();
        const z_validations = try types.CStrArray.init(alloc, validation_layers);
        defer z_validations.deinit();

        const result = instanceInit( 
            z_app_name.ptr, @intCast(version[0]), @intCast(version[1]), @intCast(version[2]), 
            z_extensions.ptr, @intCast(z_extensions.len), flags,
            z_validations.ptr, @intCast(z_validations.len)
        );
        if (result.result != vk.VK_SUCCESS) {
            return errors.instanceCreationFailed;
        }
        self.inst = result.instance;
        self.result = result;
    }

    pub fn deinit(self: *instance) void {
        instanceDeinit(&self.result);
    }
};
